src/parquet/bloom.h

```diff
@@ -0,0 +1,191 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_BLOOM_H
+#define PARQUET_BLOOM_H
+
+#include <set>
+#include <cstdint>
+
+#include "parquet/types.h"
+#include "parquet/util/memory.h"
+
+namespace parquet {
+class OutputStream;
+
+// Bloom Filter is a compact structure to indicate whether an item is not in set or
+// probably in set. Bloom class is underlying class of Bloom Filter which stores a
+// bit set represents elements set, hash strategy and bloom filter algorithm.
+
+// Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s
+// "Cache-,Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the
+// item to a tiny Bloom Filter which size fit a single cache line or smaller. This
+// implementation sets 8 bits in each tiny Bloom Filter. Tiny bloom filter are 32
+// bytes to take advantage of 32-bytes SIMD instruction.
+
+class Bloom final {
+ public:
+  // Hash strategy available for bloom filter.
+  enum HashStrategy {
+    MURMUR3_X64_128
+  };
+
+  // Bloom filter algorithm.
+  enum Algorithm {
+    BLOCK
+  };
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  static constexpr double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  static constexpr int HEADER_SIZE = 12;
+
+  // Bytes in a tiny bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // Default seed for hash function, to keep cross compatibility the seed is same as
+  // value defined in parquet-mr.
+  static constexpr int DEFAULT_SEED = 104729;
+
+  // Default maximum bloom filter size (need to discuss)
+  static constexpr int DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES = 16 * 1024 * 1024;
+
+  // The block-based algorithm needs eight odd SALT values to calculate eight indexes
+  // of bit to set, one bit in each 32-bit word.
+  static constexpr uint32_t SALT[8] = { 0x47b6137bU, 0x44974d91U, 0x8824ad5bU,
+      0xa2b7289dU, 0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U };
+
+  typedef void (*HashFunc)(const void *key, int length, uint32_t seed, void* out);
+
+ public:
+  /// Constructor of bloom filter. The range of num_bytes should be within
+  /// [BYTES_PER_FILTER_BLOCK, DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// round up/down to lower/upper bound if num_bytes is out of range. It will use
+  /// murmur3_x64_128 as its default hash function and the block based algorithm.
```

Comment:
I'm not sure it makes sense for this class to include both an implementation and the interface of a filter. Would it make more sense to define a  pure virtual Filter interface and then derive from that and override methods?

Shall we do the same for the hash function?

Reply:
The idea for using a function pointer is to treat murmur3 hash implementation as library functions,  such as [library ](https://docs.projects.genivi.org/ipc.common-api-dbus-runtime/2.0.4/MurmurHash3_8h.html) here.  So that one day if we want to switch to other hash in a c++ library, we only need to just assign the function pointer to it instead of change internal implementation.

Reply:
I'm sorry, but I'm not sure this addresses my concerns either about the Filter API or the Hash API. Pure abstract classes is a common idiom that more simply captures the intent that there is one API and multiple implementations.

The Derived classes, can, of course, call Murmur3 or whatever.

Reply:
Understood your point, will update after next commit.